ouroboros-network/src/Ouroboros/Network/PeerSelection/RootPeersDNS/DNSActions.hs

```diff
@@ -59,40 +59,55 @@ instance Exception exception => Exception (DNSorIOError exception) where
 -- Resource
 --
 
--- | Evolving resource; We use it to reinitialise the dns library if the
+-- | Evolving resource; We use it to reinitialise the DNS library if the
 -- `/etc/resolv.conf` file was modified.
 --
-newtype Resource m err a = Resource {
-    withResource :: m (Either err a, Resource m err a)
+-- Note: `constantResource` and `retryResource` are written using an approach
+-- inspired by _"The Different Aspects of Monads and Mixins"_, by Bruno C. d S.
+-- Oliveira.
+--
+newtype Resource m a = Resource {
+    withResource :: m (a, Resource m a)
   }
+  deriving Functor
+
+constantResource :: forall m a. Applicative m => a -> Resource m a
+constantResource = fix go
+  where
+    go :: (a -> Resource m a)
+       -> (a -> Resource m a)
+    go this a = Resource $ pure (a, this a)
 
--- | Like 'withResource' but retries until success.
+-- | A 'Resource` which will exhibit the given `Resource` but retry it with
+-- a given delay until a success. On first success it will reset the delays.
 --
-withResource' :: MonadDelay m
-              => Tracer m err
+retryResource :: forall m e a. MonadDelay m
+              => Tracer m e
               -> NonEmpty DiffTime
-              -- ^ delays between each re-try
-              -> Resource m err a
-              -> m (a, Resource m err a)
-withResource' tracer = go
+              -> Resource m (Either e a)
+              -> Resource m a
+retryResource tracer ds0 = fix go ds0
   where
-    dropHead :: NonEmpty a -> NonEmpty a
-    dropHead as@(_ :| [])  = as
-    dropHead (_ :| a : as) = a :| as
-
-    go !delays resource = do
-      er <- withResource resource
-      case er of
-        (Left err, resource') -> do
-          traceWith tracer err
-          threadDelay (NonEmpty.head delays)
-          withResource' tracer (dropHead delays) resource'
-        (Right r, resource') ->
-          pure (r, resource')
-
-constantResource :: Applicative m => a -> Resource m err a
-constantResource a = Resource (pure (Right a, constantResource a))
-
+    -- NOTE: this is more general recursion than what B. Oliveira is using. We
+    -- need 'super' as part of the recursion.  The reason is that we want to
+    -- modify `Resource` itself not a traversal.
```

Comment:
I am not sure if it is accurate to say that this is more general. It seems more ad-hoc to me, you are pulling the `super` out of your types instead of using it combinator style. I'd rephrase and say this is inspired in what B. Oliveira is using.

Reply:
This can fit into `Mixin` style, we're just ignoring the first argument: https://github.com/input-output-hk/ouroboros-network/pull/4707#discussion_r1376005467

Maybe more general is a wrong word as it fits into the same pattern, but it's a different way of modifying a predefined behaviour so I wouldn't say it's ad-hoc.